sdks/go/pkg/beam/io/spannerio/spanner_test.go

```diff
@@ -0,0 +1,249 @@
+				},
+				{
+					One: "one",
+					Two: 4,
+				},
+			},
+		},
+		{
+			name:     "Successfully read 1 rows",
+			database: "projects/fake-proj/instances/fake-instance/databases/fake-db-1-rows",
+			rows: []TestDto{
+				{
+					One: "one",
+					Two: 1,
+				},
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			srv, srvCleanup := newServer(t)
+			defer srvCleanup()
+
+			client, cleanup := createFakeClient(t, srv.Addr, testCase.database)
+			defer cleanup()
+
+			ddl, err := spansql.ParseDDL("",
+				`CREATE TABLE Test (
+					One STRING(20),
+					Two INT64,
+				) PRIMARY KEY (Two)`)
+			if err != nil {
+				t.Fatalf("Unable to create DDL statement for spanner test: %v", err)
+			}
+
+			err = srv.UpdateDDL(ddl)
+			if err != nil {
+				t.Fatalf("Unable to run DDL into spanner db: %v", err)
+			}
+
+			var mutations []*spanner.Mutation
+			for _, m := range testCase.rows {
+				mutation, err := spanner.InsertStruct("Test", m)
+				if err != nil {
+					t.Fatalf("Unable to create mutation to insert struct: %v", err)
+				}
+
+				mutations = append(mutations, mutation)
+			}
+
+			_, err = client.Apply(context.Background(), mutations)
+			if err != nil {
+				t.Fatalf("Applying mutations: %v", err)
+			}
+
+			p := beam.NewPipeline()
+			s := p.Root()
+			rows := query(s, "", client, "SELECT * from Test", reflect.TypeOf(TestDto{}))
+
+			passert.Count(s, rows, "", len(testCase.rows))
+			ptest.RunAndValidate(t, p)
+		})
+	}
+}
+
+func TestWrite(t *testing.T) {
+	testCases := []struct {
+		name          string
+		database      string
+		rows          []TestDto
+		expectedError bool
+	}{
+		{
+			name:     "Successfully write 4 rows",
+			database: "projects/fake-proj/instances/fake-instance/databases/fake-db-4-rows",
+			rows: []TestDto{
+				{
+					One: "one",
+					Two: 1,
+				},
+				{
+					One: "one",
+					Two: 2,
+				},
+				{
+					One: "one",
+					Two: 3,
+				},
+				{
+					One: "one",
+					Two: 4,
+				},
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			srv, srvCleanup := newServer(t)
+			defer srvCleanup()
+
+			client, cleanup := createFakeClient(t, srv.Addr, testCase.database)
+			defer cleanup()
+
+			ddl, err := spansql.ParseDDL("",
+				`CREATE TABLE Test (
+					One STRING(20),
+					Two INT64,
+				) PRIMARY KEY (Two)`)
+			if err != nil {
+				t.Fatalf("Unable to create DDL statement for spanner test: %v", err)
+			}
+
+			err = srv.UpdateDDL(ddl)
+			if err != nil {
+				t.Fatalf("Unable to run DDL into spanner db: %v", err)
+			}
+
+			p, s, col := ptest.CreateList(testCase.rows)
+
+			write(s, "", client, "Test", col)
+
+			ptest.RunAndValidate(t, p)
+
+			verifyClient, verifyClientCleanup := createFakeClient(t, srv.Addr, testCase.database)
+			defer verifyClientCleanup()
+
+			stmt := spanner.Statement{SQL: "SELECT * FROM Test"}
+			it := verifyClient.Single().Query(context.Background(), stmt)
```

Comment:
Just commenting to confirm if the plan is to pass the client into an instance of writeFn (and queryFn above) and calling ProcessFn directly, so that code gets tested?

As written, these test the basics of the fake.

I know you're busy/working on this in your spare time, so this is probably your plan when you get a moment, but I'd rather make a comment, then accidentally ignore something.

Reply:
I don't quite understand this one - can you clarify for me please? Apologies its probably just something I don't understand... I thought line 192 passing the client in to the writeFn covered this? Have I misunderstood how the writeFn instance is used inside ptest.RunAndValidate ?

Reply:
I'm sorry. I had both the new code, and stale tabs open, and then got mixed up with the verification side that I commented on here. This LGTM at thios point, aside from the merge conflicts in go.mod+go.sum  A rebase and a go mod tidy should resolve those though. 
Lets get this in!